File: LC/NC250/contains_duplicate_2.c
// problem: https://leetcode.com/problems/contains-duplicate-ii/

// time - O(N^2), space - O(1)
// Way 1: brute force approch - using nested loop to check against each entry for every i
bool containsNearbyDuplicate(int* nums, int numsSize, int k) {
  for (int i = 0; i < numsSize; i++) {
    for (int j = i + 1; j < numsSize; j++) {
      if (nums[i] == nums[j] && abs(i - j) <= k) return true;
    }
  }

  return false;    
}

File: LC/NC250/contains_duplicate_2.py
# problem: https://leetcode.com/problems/contains-duplicate-ii/

# Way 2: Using Hashtable - to record the previously visited element
# time - O(N), space - O(N)
def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
seen = {}

for i, num in enumerate(nums):
    if num in seen:
        if abs(seen[num] - i) <= k: 
            return True

    # either update the index if value exist or add a new entry to seen
    seen[num] = i

return False



# Way 3: Using Sliding window + Hashtable - remove the element from hashtable
# as the window slides and add new element to the right, then make sure the
# hashtable always has K elements
# time - O(N), space - O(K);   k - window size
def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
  window = {}

  for i, num in enumerate(nums):
    # adjust window size (remove the first entry from window)
    if i > k:
      del window[nums[i - k - 1]]

    # check if the value is present in the window
    if num in window:
      return True
    
    # finally add the new element to window
    window[num] = i

  return False